core/dom/range.js

```diff
@@ -2843,50 +2843,125 @@ CKEDITOR.dom.range = function( root ) {
+		 * @param {Boolean} [isAbsolute=false] The function will retrieve an absolute
+		 * rectangle of the element i.e. position relative to the upper-left corner of the topmost viewport.
 		 * @returns {CKEDITOR.dom.rect[]}
 		 */
 		getClientRects: ( function() {
 			if ( this.document.getSelection !== undefined ) {
-				return function() {
+				return function( isAbsolute ) {
 					// We need to create native range so we can call native getClientRects.
 					var range = document.createRange(),
-						rectArray;
+						rectList;
 
 					range.setStart( this.startContainer.$, this.startOffset );
 					range.setEnd( this.endContainer.$, this.endOffset );
+					var first,
+						textNode,
+						itemToInsertAfter;
 
-					rectArray = CKEDITOR.tools.array.map( range.getClientRects(), function( item ) {
-						return convertRect( item );
-					} );
+					rectList = range.getClientRects();
 
-					// In some cases ( eg. ranges contain only image ) IE will return empty rectList.
-					if ( !range.collapsed && !rectArray.length ) {
-						rectArray = [ getRect( this.createBookmark() ) ];
+					if ( !rectList.length ) {
+						// If native rect list is empty, use workarounds to get rects.
+
+						if ( !range.collapsed ) {
+							// In some cases ( eg. ranges contain only image ) IE will return empty rectList.
+
+							rectList = [ getRect( this.createBookmark() ) ];
+						} else if ( this.startContainer instanceof CKEDITOR.dom.element ) {
+							// If collapsed ranges are in element add textNode and return its rects.
+
+							first = this.checkStartOfBlock();
+							textNode = new CKEDITOR.dom.text( '\u200b' );
+
+							if ( first ) {
+								this.startContainer.append( textNode, true );
+							} else {
+								itemToInsertAfter = this.startContainer.getChildren().getItem( this.startOffset - 1 );
+								textNode.insertAfter( itemToInsertAfter );
+							}
+
+							// Create native collapsed ranges inside just created textNode.
+							range.setStart( textNode.$, 0 );
+							range.setEnd( textNode.$, 0 );
+
+							rectList = range.getClientRects();
+							textNode.remove();
+						} else if ( this.startContainer instanceof CKEDITOR.dom.text ) {
+							if ( this.startContainer.getText() === '' ) {
+								// In case of empty text fill it with zero width space.
+								this.startContainer.setText( '\u200b' );
+								rectList = range.getClientRects();
+
+								this.startContainer.setText( '' );
+
+							} else {
+								// If there is text node which isn't empty, but still no rects are returned use IE8 polyfill.
+								// This happens with selection at the end of line in IE.
+								rectList =  [ convertRect( getRect( this.createBookmark() ), isAbsolute, this ) ];
+							}
+						}
 					}
 
 					range.detach();
 
-					return rectArray;
+					return CKEDITOR.tools.array.map( rectList, function( item ) {
+						return convertRect( item, isAbsolute, this );
+					}, this );
 				};
 			} else {
-				return function() {
```

Comment:
Can't this version of `getClientRects` be just deleted? It looks exactly like line 2909. In such case I'd think that the whole `getClientRects` could be further simplified.

Reply:
It can be deleted only if some changes are added to code. Every call of native `range.getClientRects()` would require a check if that method exist, since IE8 doesn't support it. Possibly other checks would be needed too. Why perform all the logic on that browser when it won't do anything useful?